src/GafferScene/Parent.cpp

```diff
@@ -46,126 +54,317 @@ Parent::Parent( const std::string &name )
+	return input == mappingPlug() || isChildrenPlug( input, inPlug()->childNamesPlug()->getName() );
 }
 
 void Parent::hashBranchChildNames( const ScenePath &parentPath, const ScenePath &branchPath, const Gaffer::Context *context, IECore::MurmurHash &h ) const
 {
-	h = childPlug()->childNamesHash( branchPath );
+	if( branchPath.size() == 0 )
+	{
+		BranchCreator::hashBranchChildNames( parentPath, branchPath, context, h );
+		ScenePlug::GlobalScope s( context );
+		mappingPlug()->hash( h );
+	}
+	else
+	{
+		const ScenePlug *sourcePlug = nullptr;
+		ScenePath source = sourcePath( branchPath, &sourcePlug );
+		h = sourcePlug->childNamesHash( source );
+	}
 }
 
 IECore::ConstInternedStringVectorDataPtr Parent::computeBranchChildNames( const ScenePath &parentPath, const ScenePath &branchPath, const Gaffer::Context *context ) const
 {
-	return childPlug()->childNames( branchPath );
+	if( branchPath.size() == 0 )
+	{
+		ScenePlug::GlobalScope s( context );
+		Private::ConstChildNamesMapPtr mapping = boost::static_pointer_cast<const Private::ChildNamesMap>( mappingPlug()->getValue() );
+		return mapping->outputChildNames();
+	}
+	else
+	{
+		const ScenePlug *sourcePlug = nullptr;
+		ScenePath source = sourcePath( branchPath, &sourcePlug );
+		return sourcePlug->childNames( source );
+	}
 }
 
 bool Parent::affectsBranchSetNames( const Gaffer::Plug *input ) const
 {
-	return input == childPlug()->setNamesPlug();
+	return isChildrenPlug( input, inPlug()->setNamesPlug()->getName() );
 }
 
 void Parent::hashBranchSetNames( const ScenePath &parentPath, const Gaffer::Context *context, IECore::MurmurHash &h ) const
 {
-	assert( parentPath.size() == 0 ); // Expectation driven by `constantBranchSetNames() == true`
-	h = childPlug()->setNamesPlug()->hash();
+	BranchCreator::hashBranchSetNames( parentPath, context, h );
+	for( ScenePlugIterator it( childrenPlug() ); !it.done(); ++it )
+	{
+		(*it)->setNamesPlug()->hash( h );
+	}
 }
 
 IECore::ConstInternedStringVectorDataPtr Parent::computeBranchSetNames( const ScenePath &parentPath, const Gaffer::Context *context ) const
 {
-	assert( parentPath.size() == 0 ); // Expectation driven by `constantBranchSetNames() == true`
-	return childPlug()->setNamesPlug()->getValue();
+	InternedStringVectorDataPtr resultData = new InternedStringVectorData;
+	vector<InternedString> &result = resultData->writable();
+	for( ScenePlugIterator it( childrenPlug() ); !it.done(); ++it )
+	{
+		// This naive approach to merging set names preserves the order of the incoming names,
+		// but at the expense of using linear search. We assume that the number of sets is small
+		// enough and the InternedString comparison fast enough that this is OK.
+		ConstInternedStringVectorDataPtr inputSetNamesData = (*it)->setNamesPlug()->getValue();
+		const vector<InternedString> &inputSetNames = inputSetNamesData->readable();
+		for( vector<InternedString>::const_iterator it = inputSetNames.begin(), eIt = inputSetNames.end(); it != eIt; ++it )
```

Comment:
isnt this masking the previous definition of `it`?

Reply:
It was, but it was working because we don't need the outer `it` in the inner loop. Moved to range-for and unique names in a31e3ff.